File: backend/modules/mentors/infrastructure/repositories/mentor_repository_impl.py
"""
Mentor Repository Implementation
Реализация репозитория через SQLAlchemy
"""

from typing import List, Optional
from sqlalchemy.orm import Session

from modules.mentors.domain.entities.mentor import Mentor
from modules.mentors.domain.repositories.mentor_repository import IMentorRepository
from modules.mentors.infrastructure.database.models import MentorModel


class MentorRepository(IMentorRepository):
    """Реализация репозитория менторов"""
    
    def __init__(self, db: Session):
        self.db = db
    
    def _to_entity(self, model: MentorModel) -> Mentor:
        """Преобразовать модель БД в сущность"""
        return Mentor(
            id=model.id,
            full_name=model.full_name,
            email=model.email,
            phone=model.phone,
            photo_url=model.photo_url,
            specialization=model.specialization,
            hourly_rate=model.hourly_rate,
            start_date=model.start_date,
            status=model.status.value,
            passport_or_inn=model.passport_or_inn,
            bank_details=model.bank_details,
            notes=model.notes,
            created_at=model.created_at,
            updated_at=model.updated_at
        )
    
    def _to_model(self, entity: Mentor) -> MentorModel:
        """Преобразовать сущность в модель БД"""
        return MentorModel(
            id=entity.id,
            full_name=entity.full_name,
            email=entity.email,
            phone=entity.phone,
            photo_url=entity.photo_url,
            specialization=entity.specialization,
            hourly_rate=entity.hourly_rate,
            start_date=entity.start_date,
            status=entity.status,
            passport_or_inn=entity.passport_or_inn,
            bank_details=entity.bank_details,
            notes=entity.notes,
            created_at=entity.created_at,
            updated_at=entity.updated_at
        )
    
    def find_all(self) -> List[Mentor]:
        """Получить всех менторов"""
        models = self.db.query(MentorModel).all()
        return [self._to_entity(model) for model in models]
    
    def find_by_id(self, mentor_id: str) -> Optional[Mentor]:
        """Получить ментора по ID"""
        model = self.db.query(MentorModel).filter(MentorModel.id == mentor_id).first()
        return self._to_entity(model) if model else None
    
    def find_by_email(self, email: str) -> Optional[Mentor]:
        """Получить ментора по email"""
        model = self.db.query(MentorModel).filter(MentorModel.email == email).first()
        return self._to_entity(model) if model else None
    
    def find_by_status(self, status: str) -> List[Mentor]:
        """Получить менторов по статусу"""
        models = self.db.query(MentorModel).filter(MentorModel.status == status).all()
        return [self._to_entity(model) for model in models]
    
    def create(self, mentor: Mentor) -> Mentor:
        """Создать ментора"""
        model = self._to_model(mentor)
        self.db.add(model)
        self.db.commit()
        self.db.refresh(model)
        return self._to_entity(model)
    
    def update(self, mentor: Mentor) -> Mentor:
        """Обновить ментора"""
        model = self.db.query(MentorModel).filter(MentorModel.id == mentor.id).first()
        if not model:
            raise ValueError(f"Mentor with id {mentor.id} not found")
        
        # Обновляем поля
        model.full_name = mentor.full_name
        model.email = mentor.email
        model.phone = mentor.phone
        model.photo_url = mentor.photo_url
        model.specialization = mentor.specialization
        model.hourly_rate = mentor.hourly_rate
        model.start_date = mentor.start_date
        model.status = mentor.status
        model.passport_or_inn = mentor.passport_or_inn
        model.bank_details = mentor.bank_details
        model.notes = mentor.notes
        model.updated_at = mentor.updated_at
        
        self.db.commit()
        self.db.refresh(model)
        return self._to_entity(model)
    
    def delete(self, mentor_id: str) -> bool:
        """Удалить ментора"""
        model = self.db.query(MentorModel).filter(MentorModel.id == mentor_id).first()
        if not model:
            return False
        
        self.db.delete(model)
        self.db.commit()
        return True
    
    def search(self, query: str) -> List[Mentor]:
        """Поиск менторов"""
        query_lower = query.lower()
        models = self.db.query(MentorModel).filter(
            (MentorModel.full_name.ilike(f"%{query_lower}%")) |
            (MentorModel.email.ilike(f"%{query_lower}%")) |
            (MentorModel.specialization.ilike(f"%{query_lower}%"))
        ).all()
        return [self._to_entity(model) for model in models]

File: frontend/modules/mentors/infrastructure/repositories/mentor.repository.impl.js
/**
 * Mentor Repository Implementation
 * Реализация работы с хранилищем менторов через Backend API
 */

import api from '../../../../core/shared/api.js';
import MentorEntity from '../../domain/entities/mentor.entity.js';

class MentorRepository {
  constructor() {
    this.endpoint = '/mentors';
  }

  /**
   * Преобразовать DTO из API в Entity
   */
  _dtoToEntity(dto) {
    return MentorEntity.fromJSON({
      id: dto.id,
      fullName: dto.fullName,
      email: dto.email,
      phone: dto.phone,
      photoUrl: dto.photoUrl,
      specialization: dto.specialization,
      hourlyRate: dto.hourlyRate,
      startDate: dto.startDate,
      status: dto.status,
      passportOrInn: dto.passportOrInn,
      bankDetails: dto.bankDetails,
      notes: dto.notes,
      createdAt: dto.createdAt,
      updatedAt: dto.updatedAt
    });
  }

  /**
   * Преобразовать Entity в DTO для API
   */
  _entityToDto(entity) {
    return {
      fullName: entity.fullName,
      email: entity.email,
      phone: entity.phone,
      photoUrl: entity.photoUrl,
      specialization: entity.specialization,
      hourlyRate: entity.hourlyRate,
      startDate: entity.startDate,
      status: entity.status,
      passportOrInn: entity.passportOrInn,
      bankDetails: entity.bankDetails,
      notes: entity.notes
    };
  }

  /**
   * Получить всех менторов
   */
  async findAll() {
    try {
      const data = await api.get(this.endpoint);
      return data.map(dto => this._dtoToEntity(dto));
    } catch (error) {
      console.error('[MentorRepository] Error fetching all:', error);
      throw new Error('Не удалось загрузить список менторов');
    }
  }

  /**
   * Получить ментора по ID
   */
  async findById(id) {
    try {
      const data = await api.get(`${this.endpoint}/${id}`);
      return this._dtoToEntity(data);
    } catch (error) {
      if (error.message.includes('404')) {
        return null;
      }
      console.error('[MentorRepository] Error fetching by ID:', error);
      throw new Error('Не удалось загрузить ментора');
    }
  }

  /**
   * Получить ментора по email
   */
  async findByEmail(email) {
    const mentors = await this.findAll();
    return mentors.find(mentor => mentor.email.toLowerCase() === email.toLowerCase()) || null;
  }

  /**
   * Сохранить ментора
   */
  async save(mentor) {
    try {
      const dto = this._entityToDto(mentor);
      const data = await api.post(this.endpoint, dto);
      return this._dtoToEntity(data);
    } catch (error) {
      console.error('[MentorRepository] Error saving:', error);
      throw new Error(error.message || 'Не удалось сохранить ментора');
    }
  }

  /**
   * Обновить ментора
   */
  async update(id, data) {
    try {
      const dto = this._entityToDto(data);
      const updated = await api.put(`${this.endpoint}/${id}`, dto);
      return this._dtoToEntity(updated);
    } catch (error) {
      console.error('[MentorRepository] Error updating:', error);
      throw new Error(error.message || 'Не удалось обновить ментора');
    }
  }

  /**
   * Удалить ментора
   */
  async delete(id) {
    try {
      await api.delete(`${this.endpoint}/${id}`);
      return true;
    } catch (error) {
      if (error.message.includes('404')) {
        return false;
      }
      console.error('[MentorRepository] Error deleting:', error);
      throw new Error('Не удалось удалить ментора');
    }
  }

  /**
   * Поиск менторов
   */
  async search(query) {
    try {
      const data = await api.get(`${this.endpoint}?search=${encodeURIComponent(query)}`);
      return data.map(dto => this._dtoToEntity(dto));
    } catch (error) {
      console.error('[MentorRepository] Error searching:', error);
      throw new Error('Не удалось выполнить поиск');
    }
  }

  /**
   * Фильтр по статусу
   */
  async findByStatus(status) {
    try {
      const data = await api.get(`${this.endpoint}?status=${status}`);
      return data.map(dto => this._dtoToEntity(dto));
    } catch (error) {
      console.error('[MentorRepository] Error filtering by status:', error);
      throw new Error('Не удалось отфильтровать менторов');
    }
  }

  /**
   * Проверить уникальность email
   */
  async isEmailUnique(email, excludeId = null) {
    const mentors = await this.findAll();
    const existing = mentors.find(m => 
      m.email.toLowerCase() === email.toLowerCase() && m.id !== excludeId
    );
    return !existing;
  }
}

export default MentorRepository;

